addons/delivery_provider/models/delivery_carrier.py

```diff
@@ -0,0 +1,140 @@
+
+        :param orders: A recordset of sale orders
+        :return list: A list of floats, containing the estimated price for the shipping of the sale order
+        '''
+        if hasattr(self, '%s_get_shipping_price_from_so' % self.delivery_type):
+            return getattr(self, '%s_get_shipping_price_from_so' % self.delivery_type)(orders)
+
+    def send_shipping(self, pickings):
+        ''' Send the package to the service provider
+
+        :param pickings: A recordset of pickings
+        :return list: A list of dictionaries (one per picking) containing the exact_price of the shipping
+                      and its tracking_number
+        '''
+        if hasattr(self, '%s_send_shipping' % self.delivery_type):
+            return getattr(self, '%s_send_shipping' % self.delivery_type)(pickings)
+
+    def get_tracking_link(self, pickings):
+        ''' Ask the tracking link to the service provider
+
+        :param pickings: A recordset of pickings
+        :return list: A list of string URLs, containing the tracking links for every picking
+        '''
+        if hasattr(self, '%s_get_tracking_link' % self.delivery_type):
+            return getattr(self, '%s_get_tracking_link' % self.delivery_type)(pickings)
+
+    def cancel_shipment(self, pickings):
+        ''' Cancel a shipment
+
+        :param pickings: A recordset of pickings
+        '''
+        if hasattr(self, '%s_cancel_shipment' % self.delivery_type):
+            return getattr(self, '%s_cancel_shipment' % self.delivery_type)(pickings)
+
+
+##############################################
+#                                            #
+# Demo code: not intended for production use #
+#                                            #
+##############################################
+
+class DummyProvider(models.Model):
+    ''' This is a demo provider, intented to demonstrate how to implement a new shipping provider '''
+
+    _inherit = 'delivery.carrier'
+
+    delivery_type = fields.Selection(selection_add=[('dummy', 'Dummy Provider')])
+
+    def dummy_get_shipping_price_from_so(self, orders):
+        res = []
+        for order in orders:
+            price = 10.0 * len(order.order_line) + 5
+            res = res + [price]
+        return res
+
+    def dummy_send_shipping(self, pickings):
+        res = []
+        import random
+        for picking in pickings:
+            # For demo: random tracking number, and price can be the estimation of the SO or 1.0 more
+            shipment = {'tracking_number': picking.id + random.randint(0, 123456789),
+                        'exact_price': 10.0 * len(picking.move_lines) + 5 + random.randint(0, 1)}
+            res = res + [shipment]
+        return res
+
+    def dummy_get_tracking_link(self, pickings):
+        res = []
+        for picking in pickings:
+            res = res + [('https://www.mydummytracker.com/track?shipid=%s' % picking[0].carrier_tracking_ref)]
```

Comment:
Use an existing url.
You can use its app page (will works once merged): https://www.odoo.com/apps/modules/online/delivery_provider/